src/mockServiceWorker.js

```diff
@@ -84,150 +82,181 @@ self.addEventListener('message', async function (event) {
   }
 })
 
-self.addEventListener('fetch', function (event) {
-  const { clientId, request } = event
-  const requestId = uuidv4()
-  const requestClone = request.clone()
-  const getOriginalResponse = () => fetch(requestClone)
-
-  // Bypass navigation requests.
-  if (request.mode === 'navigate') {
-    return
-  }
+async function resolveClient(event) {
+  const client = await self.clients.get(event.clientId)
+  const allClients = await self.clients.matchAll()
+  return client.frameType === 'top-level'
+    ? client
+    : allClients.find((parentClient) => {
+        return (
+          parentClient.frameType === 'top-level' &&
+          parentClient.visibilityState === 'visible' &&
+          client.url.startsWith(parentClient.url)
```

Comment:
Here we are assuming that the first visible `top-level` that stands at the beginning of the nested client's URL is its parent window.

Reply:
This is a much better solution. Let refactor a bit of code in my test harness and then test this out. I'll get back to you soon.

Reply:
From initial looks, I'm getting an error:

```
[MSW] Failed to mock a "GET" request to "http://localhost:8080/dist/index.js": TypeError: Cannot read property 'id' of undefined
```

Let me investigate it a bit more and get back to you 😊

Reply:
`client.url.startsWith(parentClient.url)` — the client URL could be anything here:

- Parent URL for cypress is `http://localhost:8080/__/#/tests/integration/msw-requests.spec.js`
- Child URL is `http://localhost:3000`

But this URL could be anything at all

Reply:
In the case of Cypress, the parent iframe also has a type of `nested`

Reply:
Thank you for the insights, @deshiknaves!

Please, is there a way we can reproduce a Cypress iframe layout without actually using Cypress? This scenario needs to be put into an integration test for the library. Maybe you can help us out laying out the iframes structure to reproduce it?

Don't worry if you're not sure about the Cypress iframes. If you can post a reproduction repo with Cypress it'd be sufficient also, we could dig into the page structure and attempt to replicate it in an isolated test. 

> But this URL could be anything at all

Wonder how come those 2 URLs have different ports. Does it mean Cypress establishes the runner on port A and the tests themselves on port B?

Reply:
Yeah. I'll replicate this in a test. I also have some ideas on how we might be able to identify this. I'll give a few a shot tomorrow and get back.

Yes, Cypress stats an application on a port and the app you are testing might be on another port, or even a completely different domain (it doesn't have to be local).

Reply:
It's basically multiple nested iframe. Inception like — 3 levels deep.

Reply:
Sounds promising! Looking forward to seeing your test setup. We can then move it to the library to ensure this kind of nested client requests is also intercepted. Thank you!

Reply:
@kettanaito I did end up setting up a nested iframe structure in a unit test and it passed, so I had to go investigate what I looked at the other day.

So, the nesting isn't the problem, but the fact that that the URLs don't match. Removing `client.url.startsWith(parentClient.url)` will make the parent mock the requests of the child application. We could:
- Remove this check, let it mock requests that match regardless of the iframe src URL
- Pass a way for the end user to be able to pass something that will determine if the URL is valid. Maybe a RegExp, but that will have to be serialized and passed though the `event.data` to the service worker.

I can work on polishing this off if you like either of these options?

Reply:
On more investigation on the Cypress side I now have a better understanding of what the problem is in this scenario. The worker is started on a child iframe. So the `top-level` iframe isn't the one that is mocking requests. It needs to use one of the `nested` iframes instead.

If the logic is something like this:

```javascript
return client.frameType === 'top-level'
    ? client
    : allClients.find(parentClient => {
        return (
          parentClient.visibilityState === 'visible' &&
          /__cypress/.test(parentClient.url)
        )
      })
```

Then it will work. Obviously, I've hardcoded this RegExp. But if we allow the user to specify a way to determine the ancestor to use, then it will work. What do you think?

Reply:
Interesting. If the worker is registered in an iframe, then any request originating from that iframe is caught by the worker automatically (iframe client equals worker controlled client). I see room for issues only in this scenario:

```
parent
  - iframe (Cypress, MSW worker)
    - nested iframe -- request --!
```

Reply:
That's exactly the situation here. The request is in a different iframe but worker is also in a nested iframe.

Reply:
Thanks once more for the detailed investigation about how Cypress iframes are structured. This scenario should be covered in the integration tests now.